### PR TITLE
[POAE7-2681] LLVMJIT ASM Dump

### PR DESCRIPTION
**Pull request by bigPYJ1151**
### What changes were proposed in this pull request?

Add LLVMJIT ASM Dump feature.

### Why are the changes needed?

Performance evaluation.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unnecessary

### Which label does this PR belong to?

Infra

~~Pending on #251~~

Pls take a review @Ma-Jian1 @xieqi @winningsix

### CODE REVIEW

**Review by Ma-Jian1 — APPROVED** [review]

File: src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.cpp [hunk 1]
```diff
@@ -106,20 +109,40 @@ static llvm::SubtargetFeatures buildTargetFeatures(const CompilationOptions& co)
...
+  } else {
+    llvm::legacy::PassManager pass_mgr;
+    llvm::TargetMachine* tm = engine.engine->getTargetMachine();
+
+    tm->Options.MCOptions.AsmVerbose = true;
+    pass_mgr.add(llvm::createTargetTransformInfoWrapperPass(tm->getTargetIRAnalysis()));
+    tm->addPassesToEmitFile(
+        pass_mgr, file, nullptr, llvm::TargetMachine::CGFT_AssemblyFile);
+
+    pass_mgr.run(*llvm_module_);
```
Comment (Ma-Jian1) [inline src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.cpp:141]: can we just dump query function?

Comment (bigPYJ1151) [reply src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.cpp:141]: Nope, but you can found it easily at the end. For example:
```asm
        .text
        .file   "RuntimeFunctions.cpp"
        .globl  get_query_context_item_ptr # -- Begin function get_query_context_item_ptr
        .p2align        4, 0x90
        .type   get_query_context_item_ptr,@function
get_query_context_item_ptr:             # @get_query_context_item_ptr
        .cfi_startproc
# %bb.0:
        movq    (%rdi), %rax
        movq    (%rax,%rsi,8), %rax
        retq
.Lfunc_end0:
        .size   get_query_context_item_ptr, .Lfunc_end0-get_query_context_item_ptr
        .cfi_endproc
                                        # -- End function
        .globl  get_arrow_array_ptr     # -- Begin function get_arrow_array_ptr
        .p2align        4, 0x90
        .type   get_arrow_array_ptr,@function
get_arrow_array_ptr:                    # @get_arrow_array_ptr
        .cfi_startproc
# %bb.0:
        leaq    72(%rdi), %rax
        retq
.Lfunc_end1:
        .size   get_arrow_array_ptr, .Lfunc_end1-get_arrow_array_ptr
        .cfi_endproc
                                        # -- End function
        .globl  extract_arrow_array_buffer # -- Begin function extract_arrow_array_buffer
        .p2align        4, 0x90
        .type   extract_arrow_array_buffer,@function
extract_arrow_array_buffer:             # @extract_arrow_array_buffer
        .cfi_startproc
# %bb.0:
        movq    40(%rdi), %rax
        movq    (%rax,%rsi,8), %rax
        retq
.Lfunc_end2:
        .size   extract_arrow_array_buffer, .Lfunc_end2-extract_arrow_array_buffer
        .cfi_endproc
                                        # -- End function
        .globl  extract_arrow_array_child # -- Begin function extract_arrow_array_child
        .p2align        4, 0x90
        .type   extract_arrow_array_child,@function
extract_arrow_array_child:              # @extract_arrow_array_child
        .cfi_startproc
# %bb.0:
        movq    48(%rdi), %rax
        movq    (%rax,%rsi,8), %rax
        retq
.Lfunc_end3:
        .size   extract_arrow_array_child, .Lfunc_end3-extract_arrow_array_child
        .cfi_endproc
                                        # -- End function
        .globl  extract_arrow_array_len # -- Begin function extract_arrow_array_len
        .p2align        4, 0x90
        .type   extract_arrow_array_len,@function
extract_arrow_array_len:                # @extract_arrow_array_len
        .cfi_startproc
# %bb.0:
        movq    (%rdi), %rax
        retq
.Lfunc_end4:
        .size   extract_arrow_array_len, .Lfunc_end4-extract_arrow_array_len
        .cfi_endproc
                                        # -- End function
        .globl  set_arrow_array_len     # -- Begin function set_arrow_array_len
        .p2align        4, 0x90
        .type   set_arrow_array_len,@function
set_arrow_array_len:                    # @set_arrow_array_len
        .cfi_startproc
# %bb.0:
        movq    %rsi, (%rdi)
        retq
.Lfunc_end5:
        .size   set_arrow_array_len, .Lfunc_end5-set_arrow_array_len
        .cfi_endproc
                                        # -- End function
        .globl  allocate_arrow_array_buffer # -- Begin function allocate_arrow_array_buffer
        .p2align        4, 0x90
        .type   allocate_arrow_array_buffer,@function
allocate_arrow_array_buffer:            # @allocate_arrow_array_buffer
        .cfi_startproc
# %bb.0:
        movq    72(%rdi), %rdi
        movabsq $_ZN27CiderArrowArrayBufferHolder11allocBufferEmm, %rax
        jmpq    *%rax                   # TAILCALL
.Lfunc_end6:
        .size   allocate_arrow_array_buffer, .Lfunc_end6-allocate_arrow_array_buffer
        .cfi_endproc
                                        # -- End function
        .globl  query_func              # -- Begin function query_func
        .p2align        4, 0x90
        .type   query_func,@function
query_func:                             # @query_func
        .cfi_startproc
# %bb.0:                                # %.Start
        pushq   %rbp
        .cfi_def_cfa_offset 16
        pushq   %r15
        .cfi_def_cfa_offset 24
        pushq   %r14
        .cfi_def_cfa_offset 32
        pushq   %r13
        .cfi_def_cfa_offset 40
        pushq   %r12
        .cfi_def_cfa_offset 48
        pushq   %rbx
        .cfi_def_cfa_offset 56
        subq    $24, %rsp
        .cfi_def_cfa_offset 80
        .cfi_offset %rbx, -56
        .cfi_offset %r12, -48
        .cfi_offset %r13, -40
        .cfi_offset %r14, -32
        .cfi_offset %r15, -24
        .cfi_offset %rbp, -16
        movq    48(%rsi), %rax
        movq    (%rax), %rcx
        movq    40(%rcx), %rdx
        movq    8(%rdx), %r15
        movq    8(%rax), %rax
        movq    40(%rax), %rax
        movq    8(%rax), %r12
        movq    (%rcx), %rbx
        movq    (%rdi), %rax
        movq    (%rax), %r14
        movq    120(%r14), %rax
        movq    (%rax), %rbp
        leaq    (,%rbx,8), %rdx
        movq    %rdx, (%rsp)            # 8-byte Spill
        movq    72(%rbp), %rdi
        movabsq $_ZN27CiderArrowArrayBufferHolder11allocBufferEmm, %r13
        movl    $1, %esi
        callq   *%r13
        movq    %r13, %rcx
        movq    %rbp, 8(%rsp)           # 8-byte Spill
        movq    40(%rbp), %rax
        movq    8(%rax), %r13
        movq    %r14, 16(%rsp)          # 8-byte Spill
        movq    120(%r14), %rax
        movq    8(%rax), %r14
        movq    72(%r14), %rdi
        movl    $1, %esi
        movq    (%rsp), %rdx            # 8-byte Reload
        callq   *%rcx
        xorl    %eax, %eax
        movq    40(%r14), %rcx
        movq    8(%rcx), %rdx
        xorl    %ecx, %ecx
        cmpq    %rbx, %rax
        jl      .LBB7_2
        jmp     .LBB7_5
        .p2align        4, 0x90
.LBB7_4:                                # %.Loop_Update
                                        #   in Loop: Header=BB7_2 Depth=1
        addq    $1, %rax
        cmpq    %rbx, %rax
        jge     .LBB7_5
.LBB7_2:                                # %.If_Condition
                                        # =>This Inner Loop Header: Depth=1
        movq    (%r12,%rax,8), %rdi
        movq    (%r15,%rax,8), %rsi
        cmpq    %rdi, %rsi
        jge     .LBB7_4
# %bb.3:                                # %.If_True
                                        #   in Loop: Header=BB7_2 Depth=1
        leaq    (%rsi,%rdi), %rbp
        subq    %rdi, %rsi
        movq    %rbp, (%r13,%rcx,8)
        movq    %rsi, (%rdx,%rcx,8)
        addq    $1, %rcx
        jmp     .LBB7_4
.LBB7_5:                                # %.After_Loop
        movq    16(%rsp), %rax          # 8-byte Reload
        movq    %rcx, 72(%rax)
        movq    8(%rsp), %rax           # 8-byte Reload
        movq    %rcx, (%rax)
        movq    %rcx, (%r14)
        addq    $24, %rsp
        .cfi_def_cfa_offset 56
        popq    %rbx
        .cfi_def_cfa_offset 48
        popq    %r12
        .cfi_def_cfa_offset 40
        popq    %r13
        .cfi_def_cfa_offset 32
        popq    %r14
        .cfi_def_cfa_offset 24
        popq    %r15
        .cfi_def_cfa_offset 16
        popq    %rbp
        .cfi_def_cfa_offset 8
        retq
.Lfunc_end7:
        .size   query_func, .Lfunc_end7-query_func
        .cfi_endproc
                                        # -- End function

        .ident  "clang version 9.0.1 "
        .section        ".note.GNU-stack","",@progbits
```

File: src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.cpp [hunk 2]
```diff
@@ -106,20 +109,40 @@ static llvm::SubtargetFeatures buildTargetFeatures(const CompilationOptions& co)
...
                                           llvm::None,
-                                          co.aggressive_jit_compile
+                                          module_.co_.aggressive_jit_compile
                                               ? llvm::CodeGenOpt::Aggressive
                                               : llvm::CodeGenOpt::Default,
                                           true);
 }
 
+void LLVMJITEngineBuilder::dumpASM(LLVMJITEngine& engine) {
+  const std::string fname = llvm_module_->getModuleIdentifier() + ".s";
```
Comment (Ma-Jian1) [inline src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.cpp:126]: maybe we need to add some random to support dump code many times.

**Review by jikunshang — APPROVED** [review]

File: src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.cpp (hunk 2 above)
Comment (bigPYJ1151) [reply src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.cpp:126]: right, maybe we use timestamp as suffix later.

**Pull request merged**